src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,455 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+
+#include <sof/list.h>
+#include <sof/spinlock.h>
+#include <sof/audio/component.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/lib/notifier.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+#include <kernel.h>
+
+/* per-scheduler data */
+struct zephyr_ll {
+	struct list_item tasks;			/* list of ll tasks */
+	unsigned int n_tasks;			/* task counter */
+	spinlock_t lock;			/* protects the list of tasks and the counter */
+	struct ll_schedule_domain *ll_domain;	/* scheduling domain */
+};
+
+/* per-task scheduler data: so far actually unused */
+struct zephyr_ll_pdata {
+	uint64_t period;
+	bool run;
+	struct k_sem sem;
+	struct k_thread *terminator;
+};
+
+/* FIXME: would be good to use the timer, configured in the underlying domain */
+#define domain_time(sch) platform_timer_get_atomic(timer_get())
+
+/* Locking: caller should hold the domain lock */
+static void zephyr_ll_task_done(struct zephyr_ll *sch,
+				struct task *task)
+{
+	list_item_del(&task->list);
+
+	if (!sch->n_tasks) {
+		tr_info(&ll_tr, "task count underrun!");
+		k_panic();
+	}
+
+	task->state = SOF_TASK_STATE_FREE;
+
+	tr_info(&ll_tr, "task complete %p %pU", task, task->uid);
+	tr_info(&ll_tr, "num_tasks %d total_num_tasks %d",
+		sch->n_tasks, atomic_read(&sch->ll_domain->total_num_tasks));
+
+	/*
+	 * If this is the last task, domain_unregister() won't return. It is
+	 * important to decrement the task counter last before aborting the
+	 * thread.
+	 */
+	domain_unregister(sch->ll_domain, task, --sch->n_tasks);
+}
+
+/* The caller must hold the lock and possibly disable interrupts */
+static void zephyr_ll_task_insert_unlocked(struct zephyr_ll *sch, struct task *task)
+{
+	struct task *task_iter;
+	struct list_item *list;
+
+	task->state = SOF_TASK_STATE_QUEUED;
+
+	/*
+	 * Tasks are added into the list from highest to lowest priority. This
+	 * way they can then be run in the same order. Tasks with the same
+	 * priority are served on a first-come-first-serve basis
+	 */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+		if (task->priority < task_iter->priority) {
+			list_item_append(&task->list, &task_iter->list);
+			break;
+		}
+	}
+
+	/*
+	 * If the task has not been added, it means that it has the lowest
+	 * priority and should be added at the end of the list
+	 */
+	if (list == &sch->tasks)
+		list_item_append(&task->list, &sch->tasks);
+}
+
+/*
+ * Task state machine:
+ * INIT:	initialized
+ * QUEUED:	inserted into the scheduler queue
+ * RUNNING:	the scheduler is running, the task is moved to a temporary list
+ *		and then executed
+ * CANCEL:	anywhere, where the lock is not held, tasks can be cancelled
+ *		asynchronously from an arbitrary context
```

Comment:
The above text for CANCEL is not really describing the state, but rather the transition to cancel.